File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/Bithumb.java
package com.mobnetic.coinguardian.model.market;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;
import com.mobnetic.coinguardian.model.currency.Currency;
import com.mobnetic.coinguardian.model.currency.VirtualCurrency;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.LinkedHashMap;

public class Bithumb extends Market {

    private final static String NAME = "Bithumb";
    private final static String TTS_NAME = NAME;
    private final static String URL_TICKER = "https://api.bithumb.com/public/ticker/%1$s";
    private final static String URL_ORDERS = "https://api.bithumb.com/public/orderbook/%1$s?count=1";
    private final static HashMap<String, CharSequence[]> CURRENCY_PAIRS = new LinkedHashMap<>();

    static {
        CURRENCY_PAIRS.put(VirtualCurrency.BTC, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.ETH, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.ETC, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.DASH, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.LTC, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.XRP, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.BCH, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.XMR, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.ZEC, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.QTUM, new String[]{
                Currency.KRW
        });
    }

    public Bithumb() {
        super(NAME, TTS_NAME, CURRENCY_PAIRS);
    }

    @Override
    public int getNumOfRequests(CheckerInfo checkerRecord) {
        return 2;
    }

    @Override
    public String getUrl(int requestId, CheckerInfo checkerInfo) {
        if (requestId == 0) {
            return String.format(URL_TICKER, checkerInfo.getCurrencyBaseLowerCase());
        } else {
            return String.format(URL_ORDERS, checkerInfo.getCurrencyBaseLowerCase());
        }
    }

    @Override
    protected void parseTickerFromJsonObject(int requestId, JSONObject jsonObject, Ticker ticker,
                                             CheckerInfo checkerInfo) throws Exception {
        JSONObject dataObject = jsonObject.getJSONObject("data");

        if (requestId == 0) {
            ticker.vol = dataObject.getDouble("volume_1day");
            ticker.high = dataObject.getDouble("max_price");
            ticker.low = dataObject.getDouble("min_price");
            ticker.last = dataObject.getDouble("closing_price");
            ticker.timestamp = dataObject.getLong("date");
        } else {
            ticker.bid = getFirstPriceFromOrder(dataObject, "bids");
            ticker.ask = getFirstPriceFromOrder(dataObject, "asks");
        }
    }

    @Override
    protected String parseErrorFromJsonObject(int requestId, JSONObject jsonObject,
                                              CheckerInfo checkerInfo) throws Exception {
        return jsonObject.getString("message");
    }

    private double getFirstPriceFromOrder(JSONObject jsonObject, String key) throws Exception {
        JSONArray array = jsonObject.getJSONArray(key);

        if (array.length() == 0) {
            return Ticker.NO_DATA;
        }

        JSONObject first = array.getJSONObject(0);
        return first.getDouble("price");
    }
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/Bithumb.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.Market
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import com.aneonex.bitcoinchecker.datamodule.model.currency.Currency
import com.aneonex.bitcoinchecker.datamodule.model.currency.VirtualCurrency
import com.aneonex.bitcoinchecker.datamodule.model.currency.CurrencyPairsMap
import org.json.JSONObject

class Bithumb : Market(NAME, TTS_NAME, CURRENCY_PAIRS) {
    companion object {
        private const val NAME = "Bithumb"
        private const val TTS_NAME = NAME
        private const val URL_TICKER = "https://api.bithumb.com/public/ticker/%1\$s"
        private const val URL_ORDERS = "https://api.bithumb.com/public/orderbook/%1\$s?count=1"
        private val CURRENCY_PAIRS: CurrencyPairsMap = CurrencyPairsMap()

        init {
            val marketAll = arrayOf(Currency.KRW, VirtualCurrency.BTC)
            val marketKrw = arrayOf(Currency.KRW)
            val marketBtc = arrayOf(VirtualCurrency.BTC)

            CURRENCY_PAIRS[VirtualCurrency.BTC] = marketKrw
            CURRENCY_PAIRS[VirtualCurrency.ETH] = marketAll
            CURRENCY_PAIRS[VirtualCurrency.ETC] = marketAll
            CURRENCY_PAIRS[VirtualCurrency.LTC] = marketAll
            CURRENCY_PAIRS[VirtualCurrency.XRP] = marketAll
            CURRENCY_PAIRS[VirtualCurrency.BCH] = marketAll
            CURRENCY_PAIRS[VirtualCurrency.DOGE] = marketAll
            CURRENCY_PAIRS[VirtualCurrency.SOL] = marketAll

            CURRENCY_PAIRS[VirtualCurrency.QTUM] = marketKrw
            CURRENCY_PAIRS[VirtualCurrency.CON] = marketKrw
            CURRENCY_PAIRS[VirtualCurrency.TRX] = marketKrw

            CURRENCY_PAIRS["LN"] = marketBtc
            CURRENCY_PAIRS["IBP"] = marketBtc
            CURRENCY_PAIRS["BFC"] = marketAll

            CURRENCY_PAIRS[VirtualCurrency.EOS] = marketAll
            CURRENCY_PAIRS[VirtualCurrency.LINK] = marketKrw

            CURRENCY_PAIRS["WEMIX"] = marketKrw
            CURRENCY_PAIRS["BORA"] = marketKrw
            CURRENCY_PAIRS["ASM"] = marketKrw
            CURRENCY_PAIRS["SAND"] = marketKrw
            CURRENCY_PAIRS["MANA"] = marketKrw
            CURRENCY_PAIRS["YFI"] = marketKrw
            CURRENCY_PAIRS["DVI"] = marketKrw
            CURRENCY_PAIRS["ETC"] = marketKrw
            CURRENCY_PAIRS["DOT"] = marketKrw
            CURRENCY_PAIRS["LUNA"] = marketKrw
        }
    }

    override fun getNumOfRequests(checkerInfo: CheckerInfo?): Int {
        return 2
    }

    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        val pairId = "${checkerInfo.currencyBaseLowerCase}_${checkerInfo.currencyCounterLowerCase}"

        return if (requestId == 0) {
            String.format(URL_TICKER, pairId)
        } else {
            String.format(URL_ORDERS, pairId)
        }
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker,
                                           checkerInfo: CheckerInfo) {
        val dataObject = jsonObject.getJSONObject("data")
        if (requestId == 0) {
            ticker.vol = dataObject.getDouble("units_traded_24H")
            ticker.high = dataObject.getDouble("max_price")
            ticker.low = dataObject.getDouble("min_price")
            ticker.last = dataObject.getDouble("closing_price")
            ticker.timestamp = dataObject.getLong("date")
        } else {
            ticker.bid = getFirstPriceFromOrder(dataObject, "bids")
            ticker.ask = getFirstPriceFromOrder(dataObject, "asks")
        }
    }

    @Throws(Exception::class)
    override fun parseErrorFromJsonObject(requestId: Int, jsonObject: JSONObject,
                                          checkerInfo: CheckerInfo?): String? {
        return jsonObject.getString("message")
    }

    @Throws(Exception::class)
    private fun getFirstPriceFromOrder(jsonObject: JSONObject, key: String): Double {
        val array = jsonObject.getJSONArray(key)
        if (array.length() == 0) {
            return Ticker.NO_DATA.toDouble()
        }
        val first = array.getJSONObject(0)
        return first.getDouble("price")
    }
}